File: 3286. Find a Safe Walk Through a Grid/solution.py
class Solution:
    def findSafeWalk(self, grid: List[List[int]], health: int) -> bool:
        m, n = len(grid), len(grid[0])

        dist = [[-1] * n for _ in range(m)]
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        min_heap = [ (grid[0][0], 0, 0) ]

        while min_heap:
            cost, x, y = heapq.heappop(min_heap)

            if dist[x][y] >= 0:
                continue
            
            dist[x][y] = cost

            for dx, dy in dirs:
                nx, ny = x + dx, y + dy

                if 0 <= nx < m and 0 <= ny < n and dist[nx][ny] == -1:
                    heapq.heappush(min_heap, (cost + grid[nx][ny], nx, ny))
        
        return dist[m-1][n-1] < health

File: 3286. Find a Safe Walk Through a Grid/solution.swift
class Solution {
    func findSafeWalk(_ grid: [[Int]], _ health: Int) -> Bool {
        let m = grid.count
        let n = grid[0].count

        let directions: [(Int, Int)] = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        var bestCost: [[Int]] = Array(repeating: Array(repeating: 10_000, count: n), count: m)

        var heap = Heap<Node>()

        heap.insert(Node(r: 0, c: 0, cost: grid[0][0]))

        while let node = heap.popMin() {
            let r = node.r
            let c = node.c
            let cost = node.cost

            guard cost < bestCost[r][c] else { continue }

            bestCost[r][c] = cost

            if r == m - 1, c == n - 1 {
                break
            }

            for (dr, dc) in directions {
                let nr = r + dr
                let nc = c + dc

                guard nr >= 0, nr < m, nc >= 0, nc < n else { continue }

                let candidate = cost + grid[nr][nc]

                if candidate < bestCost[nr][nc] {
                    heap.insert(Node(r: nr, c: nc, cost: candidate))
                }
            }
        }

        return health - bestCost[m-1][n-1] >= 1
    }

    struct Node: Comparable {
        var r: Int
        var c: Int

        var cost: Int

        static func < (_ lhs: Node, _ rhs: Node) -> Bool {
            lhs.cost < rhs.cost
        }
    }
}
